Mathlib/LinearAlgebra/LinearIndependent.lean

```diff
@@ -131,41 +133,635 @@ theorem linearIndependent_iff_injective_linearCombination :
+
+lemma LinearIndependent.eq_of_pair {x y : M} (h : LinearIndependent R ![x, y])
+    {s t s' t' : R} (h' : s • x + t • y = s' • x + t' • y) : s = s' ∧ t = t' :=
+  pair_iffₛ.mp h _ _ _ _ h'
+
+lemma LinearIndependent.eq_zero_of_pair' {x y : M} (h : LinearIndependent R ![x, y])
+    {s t : R} (h' : s • x = t • y) : s = 0 ∧ t = 0 := by
+  suffices H : s = 0 ∧ 0 = t from ⟨H.1, H.2.symm⟩
+  exact h.eq_of_pair (by simpa using h')
+
+lemma LinearIndependent.eq_zero_of_pair {x y : M} (h : LinearIndependent R ![x, y])
+    {s t : R} (h' : s • x + t • y = 0) : s = 0 ∧ t = 0 := by
+  replace h := @h (.single 0 s + .single 1 t) 0 ?_
+  · exact ⟨by simpa using congr($h 0), by simpa using congr($h 1)⟩
+  simpa
+
+/-- A family is linearly independent if and only if all of its finite subfamily is
+linearly independent. -/
+theorem linearIndependent_iff_finset_linearIndependent :
+    LinearIndependent R v ↔ ∀ (s : Finset ι), LinearIndependent R (v ∘ (Subtype.val : s → ι)) :=
+  ⟨fun H _ ↦ H.comp _ Subtype.val_injective, fun H ↦ linearIndependent_iff'ₛ.2 fun s f g eq i hi ↦
+    Fintype.linearIndependent_iffₛ.1 (H s) (f ∘ Subtype.val) (g ∘ Subtype.val)
+      (by simpa only [← s.sum_coe_sort] using eq) ⟨i, hi⟩⟩
+
+theorem LinearIndependent.coe_range (i : LinearIndependent R v) :
+    LinearIndependent R ((↑) : range v → M) := by simpa using i.comp _ (rangeSplitting_injective v)
+
+/-- If `v` is an injective family of vectors such that `f ∘ v` is linearly independent, then `v`
+    spans a submodule disjoint from the kernel of `f` -/
+theorem Submodule.range_ker_disjoint {f : M →ₗ[R] M'}
+    (hv : LinearIndependent R (f ∘ v)) :
+    Disjoint (span R (range v)) (LinearMap.ker f) := by
+  rw [LinearIndependent, Finsupp.linearCombination_linear_comp] at hv
+  rw [disjoint_iff_inf_le, ← Set.image_univ, Finsupp.span_image_eq_map_linearCombination,
+    map_inf_eq_map_inf_comap, (LinearMap.ker_comp _ _).symm.trans
+      (LinearMap.ker_eq_bot_of_injective hv), inf_bot_eq, map_bot]
+
+/-- If the image of a family of vectors under a linear map is linearly independent, then so is
+the original family. -/
+theorem LinearIndependent.of_comp (f : M →ₗ[R] M') (hfv : LinearIndependent R (f ∘ v)) :
+    LinearIndependent R v := by
+  rw [LinearIndependent, Finsupp.linearCombination_linear_comp, LinearMap.coe_comp] at hfv
+  exact hfv.of_comp
+
+/-- If `f` is an injective linear map, then the family `f ∘ v` is linearly independent
+if and only if the family `v` is linearly independent. -/
+protected theorem LinearMap.linearIndependent_iffₛ (f : M →ₗ[R] M') (hf_inj : Injective f) :
+    LinearIndependent R (f ∘ v) ↔ LinearIndependent R v := by
+  simp_rw [LinearIndependent, Finsupp.linearCombination_linear_comp, coe_comp, hf_inj.of_comp_iff]
+
+/-- An injective linear map sends linearly independent families of vectors to linearly independent
+families of vectors. See also `LinearIndependent.map` for a more general statement. -/
```

Comment:
The generality is now incomparable! My draft at #18503 restates `map` as
```lean
theorem LinearIndependent.map (hv : LinearIndependent R v) {f : M →ₗ[R] M'}
    (hf : Set.InjOn f (span R (range v))):
    LinearIndependent R (f ∘ v) := by
```
which keeps the support for semirings

Reply:
This is nice, I'd certainly adopt it.

Reply:
Perhaps to make things reviewable, we should:
1. Pre-emptively restate the theorems such that they _could_ be generalized to semirings (a spin off PR)
2. Move all the theorems around without changing their statement (this PR)

Otherwise I think we might want to spin up leaff here...

Reply:
Is it that hard to review? Declaration diff shows no decls are removed. New decls with a s subscript are straightforward generalization of the ring version without the s.

Reply:
I think I've incorporated all generalizations from your PR, and I added you as a co-author.

Reply:
If you're worried about the accuracy of docstrings:

(1) No decls are removed; for the existing decls that are moved, the docstrings are moved together.  Since the statements don't change I think the docstrings probably also need not change.
(2) Most new decls in the file LinearIndependent are modified version of existing decls, and carry the docstrings (whenever they exist) from the existing decls. I have now checked all docstrings of new decls and modified them to match the (Semiring-specific) new statements.

Reply:
> Is it that hard to review? Declaration diff shows no decls are removed. New decls with a s subscript are straightforward generalization of the ring version without the s.

I agree with @eric-wieser that it would be nice to have leaff working on this. But afaik leaff isn't really ready for that yet, right?

Given that no decls were removed, I suggest we just move forward with this.

bors merge